ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
-          throw RequestOffBufferBounds(bufferSnapshot.vector.last._1, endInclusive)
-        } else if (bufferSnapshot.vector.isEmpty)
-          Empty
-        else {
-          val Seq(bufferStartInclusiveIdx, bufferEndExclusiveIdx) =
-            Seq(startExclusive, endInclusive)
-              .map(bufferSnapshot.vector.searchBy(_, _._1))
-              .map {
-                case InsertionPoint(insertionPoint) => insertionPoint
-                case Found(foundIndex) => foundIndex + 1
-              }
-
-          val vectorSlice =
-            bufferSnapshot.vector.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
-
-          if (bufferStartInclusiveIdx == 0) Prefix(vectorSlice)
-          else Inclusive(vectorSlice)
+          throw RequestOffBufferBounds(bufferSnapshot.rangeEnd.get, endInclusive)
+        } else {
+          val bufferStartInclusiveIdx = searchByOffset(vectorSnapshot, startExclusive)
+          val bufferEndExclusiveIdx = searchByOffset(vectorSnapshot, endInclusive)
+          val bufferSlice = vectorSnapshot.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
+
+          if (bufferStartInclusiveIdx == 0) {
+            // The buffer is potentially a suffix of the fetched window
+            if (bufferSlice.isEmpty) {
+              // If the slice is empty, the entire window is before the buffer bounds
+              Suffix(endInclusive, Vector.empty, None)
+            } else {
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice.tail, filter)
+              // We "waste" the first element of the slice in order to provide
+              // a buffered start exclusive offset that can be used in the caller as an endInclusive
+              // for back-filling the window elements that are before the buffer's bounds.
+              Suffix(bufferSlice.head._1, filteredSlice, continueFrom)
+            }
+          } else {
+            val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice, filter)
+            Inclusive(filteredSlice, continueFrom)
+          }
         }
       },
     )
 
+  private def filterAndChunkSlice[FILTER_RESULT](
+      bufferSlice: Vector[(Offset, ENTRY)],
+      filter: ENTRY => Option[FILTER_RESULT],
+  ) = {
+    var takenFiltered = 0
+    var lastScannedOffset = Offset.beforeBegin
+
+    val it = bufferSlice.iterator
+    val resultBuilder = Vector.newBuilder[(Offset, FILTER_RESULT)]
+
+    while (it.hasNext && takenFiltered < maxBufferedChunkSize) {
+      val (offset, entry) = it.next()
+      lastScannedOffset = offset
+
+      filter(entry).foreach { filtered =>
+        takenFiltered += 1
+        resultBuilder.addOne(offset -> filtered)
+      }
+    }
+
+    val filteredEntries = resultBuilder.result()
+    sliceSizeHistogram.update(filteredEntries.size)
+    (filteredEntries, Option.when(it.hasNext)(lastScannedOffset))
+  }
+
+  private def searchByOffset(buffer: Vector[(Offset, ENTRY)], offset: Offset) =
```

Comment:
As we are adding 1 maybe a better name for this would be `indexAfterOffset`

Reply:
Good idea 👍